docs/JacobianCheck.md

```diff
@@ -0,0 +1,58 @@
+# Jacobian Correctness Verification
+
+ADBench needs not only to measure the time different tools take to compute the Jacobians of objective functions, but also to check, whether produced Jacobians are accurate. This document describes, how the accuracy of the Jacobians is verified.
+
+## Source of the Ground Truth
+
+Some of our tests (most notably, those related to bundle adjustments) result in very large Jacobians. In fact, a complete set of correct Jacobians for all our tests in tsv format consumes more than 20 GB of storage. As such, it is not feasible to check such a set into the repository. So instead, we designate one of the _testing modules_ to be _golden module_ - the source of ground truth, against output of which Jacobians produced by all the other _testing modules_ will be checked.
+
+The accuracy of the results produced by the _golden module_ is verified separately.
+
+We designate the module `Manual`, which computes the Jacobians using the manually derived formulae, as the _golden module_. We do so, because
+- It is fast enough not to timeout, so when some other module produces a Jacobian, we would have a Jacobian, we consider accurate, to compare it against
+- We have the complete control over this module, so if we ever encounter a bug in it, it will be within our power to fix it
+
+Below we will refer to Jacobians produced by the _golden module_ as _golden Jacobians_.
+
+## Comparing Jacobians
+
+We consider a Jacobian produced by a _testing module_ accurate when its dimensions are equal to those of the _golden Jacobian_, and when all its elements are _near_ the corresponding elements of the _golden Jacobian_.
+
+We say, that two floating-point numbers are _near_, when the difference between them, defined as
+
+<table>
+  <tr>
+    <td>
+        <code>
+            ρ(x, y) = |x - y| / max(1, |x| + |y|)
+        </code>
+    </td>
+    <td>(1)</td>
+  </tr>
+</table>
+
+is smaller than the given _tolerance_. Note, that the formula (1) produces the absolute difference between `x` and `y`, when both lie in the vicinity of zero, and the relative one otherwise, which is a suitable way of comparing floating-point numbers.
+
+We generally use the tolerance of `1e-6` to allow for rounding errors during both the computation and the output. This number may be adjusted for specific tests, e.g. it can be increased when the _testing module_ uses single-precision arithmetic.
+
```

Comment:
1e-6 is quite large for double -- have you tried 1e-8?

Reply:
No I didn't. I picked `1e-6`, because existing python tools produced only 6 decimal digits. We'll probably change that, so we might as well decrease the default tolerance. I'll run eigen tests against tolerance of `1e-8` overnight and see if there're any problems.

I can say right away, that finite differences are less precise, though. But that's the reason we can have separate tolerances for different tools.

Reply:
Yes, finite will always be less precise, but assuming Manual is correct, then |Tool - Manual| should be much better than finite differences.

Reply:
Agreed. So eigen and eigenvector are fine with tolerance of `1e-8`. Let that be default. I set the tolerance for finite to `1e-5` so that it wouldn't produce too many errors.